TEST_APPS/device/nfcapp/nfccommands.cpp

```diff
@@ -0,0 +1,203 @@
+/*
+ * Copyright (c) 2018 ARM Limited. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ * Licensed under the Apache License, Version 2.0 (the License); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an AS IS BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <string>
+#include <vector>
+#include <stdlib.h>
+#include "mbed.h"
+#include "mbed_events.h"
+#include "mbed-client-cli/ns_cmdline.h"
+#include "rtos\Thread.h"
+
+#include "nfctestshim.h"
+#include "nfcprocess.h"
+#include "nfccommands.h"
+
+events::EventQueue nfcQueue;
+Thread nfcThread;
+NFCTestShim * pNFC_Test_Shim = NULL;
+
+
+
+NFCTestShim* new_testshim() {
+#if MBED_CONF_NFCEEPROM
+    mbed::nfc::NFCEEPROMDriver& eeprom_driver = get_eeprom_driver(nfcQueue);
+
+    return ( (NFCTestShim *)(new NFCProcessEEPROM(nfcQueue, eeprom_driver)) );
+#else
+    return( (NFCTestShim *)(new NFCProcessController(nfcQueue)) );
+#endif // EEPROM
+
+}
+
+void nfcRoutine() {
+    nfcQueue.dispatch_forever();
+}
+
+HandleTestCommand::HandleTestCommand() {
+    osStatus status = nfcThread.start(callback(&nfcRoutine) );
+    MBED_ASSERT(status == osOK);
+}
+
+int HandleTestCommand::cmd_set_last_nfc_error(int argc, char *argv[]){
+    if (argc <= 1){
+        cmd_printf("setlastnfcerror() invalid parameter(s)\r\n");
+        return(CMDLINE_RETCODE_INVALID_PARAMETERS);
+    } else {
+        int value = strtol(argv[1], NULL, 10);
+        nfcQueue.call(NFCTestShim::on_set_last_nfc_error, value);
+    }
+    return(CMDLINE_RETCODE_EXCUTING_CONTINUE);
+}
+
+
+int HandleTestCommand::cmd_init_nfc(int argc, char *argv[]){
+
+	if (pNFC_Test_Shim) {
+		cmd_printf("WARN init called again!\r\n"); // only legal here, if eeprom driver stops talking
+	} else {
+	    pNFC_Test_Shim = new_testshim();
+	}
+	nfcQueue.call(pNFC_Test_Shim, &NFCTestShim::on_init);
+	return(CMDLINE_RETCODE_EXCUTING_CONTINUE);
+}
+
+int HandleTestCommand::cmd_read_message(int argc, char *argv[]) {
+    nfcQueue.call(pNFC_Test_Shim, &NFCTestShim::on_read_nfceeprom);
+
+    return(CMDLINE_RETCODE_EXCUTING_CONTINUE);
+}
+
+int HandleTestCommand::cmd_set_smartposter(int argc, char *argv[]) {
+    if (argc <= 1){
+        cmd_printf("setlastnfcerror() invalid parameter(s)\r\n");
+        return(CMDLINE_RETCODE_INVALID_PARAMETERS);
+    } else {
+        // parse arg and queue it up
+        char * uri = (char*)malloc(MAX_URL_LENGTH+1);
```

Comment:
you can malloc and copy the correct length

Reply:
Good spot, cleaner to copy just the string.